### ISSUE TITLE
Sample object - should resolution and footprint be `Sample` or `RunAngle` attributes?

### ISSUE THREAD

**Issue by acaruana2009** [welbournR/IBEX_NR_Scripts#2]
@welbournR @maxskoda I am currently working through the `RunAngle` logic and various methods, and I am thinking that the `Sample` object should include a `sample_length` - i.e. length along the beam, to deal with over-illumination calculations in the resolution calculator.

Also, do we want resolution and footprint to be associated with a sample? or a measurement? My feelings are that these are measurement parameters, i.e. you may want to measure the same sample with a few different resolutions (say if it is thick and you want to optimise for a longer experiment).

Currently, you would need to define two identical samples and pass them through to the `RunAngle` call. looking something like:

```
sample_1_res_0p03 = samp.SampleGenerator(
                                translation=0,
                                height2_offset=0,
                                phi_offset=0,
                                psi_offset=0,
                                height_offset=0,
                                resolution=0.03,
                                footprint=50,
                                title="Block 1",
                                subtitle="Sample 1"
                                )
sample_1_res_0p02 = samp.SampleGenerator(
                                translation=0,
                                height2_offset=0,
                                phi_offset=0,
                                psi_offset=0,
                                height_offset=0,
                                resolution=0.02,
                                footprint=50,
                                title="Block 1",
                                subtitle="Sample 1"
                                )
...

# set up measurements
sample_1_measure_0p03 = ra.RunAngle(width_slits=width_slits, sample=sample_1_res_0p03)
sample_1_measure_0p02 = ra.RunAngle(width_slits=width_slits, sample=sample_1_res_0p02)

# make function calls simple
run_sample_1_0p03 = sample_1_measure_0p03.run_angle
run_sample_1_0p02 = sample_1_measure_0p02.run_angle

# run script

def run_script():
    run_sample_1_0p03(angle=0.5)
    run_sample_1_0p02(angle=0.5)
```

By putting it into the measurement you would save the extra sample definition (we can make the rest of the script simpler as well, but that is not what I would like to discuss in this issue):

```
sample_1 = samp.SampleGenerator(
                                translation=0,
                                height2_offset=0,
                                phi_offset=0,
                                psi_offset=0,
                                height_offset=0,
                                title="Block 1",
                                subtitle="Sample 1",
                                sample_length=80
                                )
...

# set up measurements
sample_1_measure_0p03 = ra.RunAngle(width_slits=width_slits, sample=sample_1, resolution=0.03, footprint=50)
sample_1_measure_0p02 = ra.RunAngle(width_slits=width_slits, sample=sample_1, resolution=0.02, footprint=50)

# make function calls simple
run_sample_1_0p03 = sample_1_measure_0p03.run_angle
run_sample_1_0p02 = sample_1_measure_0p02.run_angle

# run script

def run_script():
    run_sample_1_0p03(angle=0.5)
    run_sample_1_0p02(angle=0.5)
```

Thoughts?

**Comment by maxskoda** [comment]
Hi Andrew,

For the sample length, I agree - we should include it, as well as some
other attributes and make the sample class iterable:
from dataclasses import dataclass, field

class IterRegistry(type):
    def __iter__(cls):
        return iter(cls._registry)

@dataclass
class NRsample(object):
    __metaclass__ = IterRegistry
    _registry = []

    title: str
    translation: float
    primary_height: float
    secondary_height: float
    primary_height_name: str
    phi_offset: float
    psi: float
    footprint: float
    resolution: float
    switch_pos: int
    sample_length: int
    hgaps: dict = field(default_factory=dict)

    def __post_init__(self):
        self._registry.append(self)
##################################I

Regarding footprint and resolution, I see what you mean, but I'm not sure I
agree. I think we tried to sort attributes into either the runAngle or the
sample object based on how often we expect those to change.
For the footprint and resolution, when we need to adjust it, we either
change it in the original sample definition (not define another sample), or
you can simply do
sample1.footprint = 25
sample1.resolution = 0.02
run_angle(sample1, ...)

If you want to change it inline so to speak.
I think what you propose is fine, but it adds a lot of lines and makes the
names longer etc and if you then want to change something, you need to
modify many lines of code...I think you can still choose to do this, but I
personally would rather not have this as standard, because it doesn't
reflect how we operate in most cases: Surely if you notice your settings
and not appropriate, for the next script you simply change the settings -
at some point you will reach the optimal setup?
Happy to discuss further.

Thanks,

Max

**Comment by acaruana2009** [comment]
To clarify, the `run_script()` function is ultimately the script part while all the rest is defined at the top of the script file - and only needs to be set once, in principle. I added in the extra lines to simplify the function calls as you preferred for the actual measurement call to be simple, i.e. without ".". This aspect I think we should discuss in another issue.

As for where to include the attributes of footprint and resolution, my feeling was that these were more measurement settings rather than sample settings. However, the code could be written as such to allow for both, where one is an override of the other for example:

```
sample_1 = samp.SampleGenerator(
                                translation=0,
                                height2_offset=0,
                                phi_offset=0,
                                psi_offset=0,
                                height_offset=0,
                                title="Block 1",
                                subtitle="Sample 1",
                                footprint=50,
                                resolution=0.03,
                                sample_length=80
                                )
...

# set up measurements
# setup using resolution from sample definition:
sample_1_measure = ra.RunAngle(width_slits=width_slits, sample=sample_1)
# setup where resolution and footprint are overridden at the call to RunAngle
sample_1_measure_0p02 = ra.RunAngle(width_slits=width_slits, sample=sample_1, resolution=0.02, footprint=50)

# make function calls simple - this is only here to remove the "."
run_sample_1 = sample_1_measure.run_angle
run_sample_1_0p02 = sample_1_measure_0p02.run_angle

# run script

def run_script():
    # start of script
    run_sample_1(angle=0.5)
    run_sample_1(angle=1.5)
    run_sample_1(angle=2.3)
    
    # or to run the other setting
    run_sample_1_0p02(angle=0.5)
```

You could modify, as you suggested, in-line, but then you would need to make sure the transmissions were taken before you modified these settings. The idea I had about linking these together as each measurement setup being an instance of the `RunAngle` class was that you could call the `transmission` method for that instance and only need to pass the angle and the runtime (in uamps, frames etc...) and it would do the correct transmission setup for that measurement setup and already have the footprint and resolution passed to it. 

It is a little extra setup at the start of the script, but once that is done, the main part of the script is very simple indeed.

**Comment by maxskoda** [comment]
Hi Andrew,

I think we probably need to talk about this live, as I'm getting a bit
confused as to what we are trying to achieve here...If I understand
correctly, the code you are suggesting involves some wrappers for the
"run_angle" method that we already have. If that's correct, then I have no
objection, as it would not preclude us from using the "run_angle" as we did
before.
My main concern is that it all gets too convoluted and we're trying to
accommodate functionality for unusual scripts (though I appreciate, it may
be more common on Polref).
Specifically, I didn't understand your comment "you would need to make sure
the transmissions were taken before you modified these settings". While we
try to take transmissions before the sample runs, this is by no means
always the case. Often we use old transmission to start with and then
script tran runs over night, so I don't see this as a problem. Having the
transmission linked to the sample is certainly something we want to do and
have talked about, though it is not entirely clear that we can do that -
we'd need to have some sort of mapping from sample slits to transmission
slits. But we should certainly aim for that.
Also I feel we are digressing from the main objective, which is to find
common ground among all instruments so that we can run the same script.
I would suggest we focus on that first and then we can refine these
specific things.
Let's also wait and hear Becky's thoughts.
Let's talk
about
this at our next meeting.

Cheers,

Max

**Comment by welbournR** [comment]
Hi Both,

I have tried to follow through the discussion above and to summarise think it comes down to:
- I think we are all agreed that the sample length should be included in the sample object.
- There is then some discussion as to whether the resolution/footprint should be a measurement or a sample setting.

In regards to the second point, if I've understood correctly, this measurement layer seems to be extra on top of what we have had setup previously on INTER. We have typically had the sample settings, some instrument defaults and then the RunAngle function. For the majority of measurements (~80-90% User experiments, perhaps lower during commissioning time), the RunAngle call is then very simple as the settings are all previously defined. On the occasions of a more 'difficult' setup we have then changed the relevant settings either within the RunAngle call (e.g. a slit value) or as an additional line to set a new sample setting ahead of the RunAngle call (e.g. a translation or resolution). This approach has worked well for INTER in isolation. The question I think is would this work for POLREF or do we need the extra measurement layer? And if we have the extra layer what else would be included in this?

I think it is important that the RunAngle setup has the ability to overwrite any of the settings but with an aim to have the call as simple as possible for the majority of measurements. I think this should be possible without the extra measurement layer BUT if POLREF need to overwrite settings much more regularly than we do on INTER then perhaps this layer is needed.

In regards to the resolution/footprint/sample_length in particular, just a couple of extra points:
- we need to include some extra checks and balances to ensure that non-reasonable values are caught. I think this is already on our to-do list.
- linked to the above, I see this as a setting that the local contact would help sort at the start of the experiment and then wouldn't expect the user to be changing (except in exceptional circumstances). Is that how you view it?
- the labelling of resolution/footprint may be misleading. This is in effect the incident resolution and footprint onto the sample, rather than necessarily being what comes out in the final data. It is a tool for ensuring slits scale but are set to/labelled with something more meaningful than just the values for the lowest angle. Is this a helpful approach for POLREF experiments or do we need to set it up with different values? We can look at including any supermirror overillumination into the calculation if that would help?

Thanks,
Becky
